src/dbnode/storage/index/block.go

```diff
@@ -443,63 +444,84 @@ func (b *block) queryWithSpan(
 	opts QueryOptions,
 	results DocumentResults,
 ) (bool, error) {
+	iter, err := b.QueryIter(ctx, query)
+	if err != nil {
+		return false, err
+	}
+	iterCloser := safeCloser{closable: iter}
+	defer func() {
+		_ = iterCloser.Close()
+		b.metrics.queryDocsMatched.RecordValue(float64(results.TotalDocsCount()))
+		b.metrics.querySeriesMatched.RecordValue(float64(results.Size()))
+	}()
+
+	if err := b.QueryWithIter(ctx, opts, iter, results, math.MaxInt64); err != nil {
+		return false, err
+	}
+
+	if err := iterCloser.Close(); err != nil {
```

Comment:
Any issue with `iterCloser.Close()` being called twice (once here and the defer above)? Do we need to do that?

Reply:
i'm just preserving the existing behavior. IMO it's probably silly, but there isn't an issue. note the use of the `safeCloser` which wraps the `Closer` with a idempotent check, so it's safe to eagerly close and rely on the defer to close as well.